src/shared/qmlpath.cpp

```diff
@@ -0,0 +1,287 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "qmlpath.h"
+
+#include <QQmlApplicationEngine>
+#include <QQuickItem>
+#include <QQuickWindow>
+#include <QRegularExpression>
+#include <QTest>
+
+QmlPath::QmlPath(QObject* parent) : QObject(parent) {}
+
+bool QmlPath::parse(const QString& obj) {
+  if (obj.isEmpty()) {
+    return false;
+  }
+
+  qsizetype size = obj.size();
+  const QChar* input = obj.constData();
+
+  QList<Data> blocks;
+
+  while (true) {
```

Comment:
This took me a while to get my head around to check for correctness, but I think it's good now.